### PR TITLE
Save user response to report form on Datastore

### LINKED ISSUE #1: Store markers generated by users
Store the location, time, title, and description on datastore specified by user in the form

### PR DESCRIPTION
**Pull request by jiminbyun**
Closes #1 
It reads the form and save it to datastore as report, which has following attributes: title, longitude, latitude, timestamp, incidentType, description. Map was used between parameter name and html id, and between parameter name and default value because first there are a lot of fields, and also it will be easier to add extra field for images. Also, instead of using button type as submit for submit button, I used regular button that fetches with method POST on click to avoid reloading the whole page. 

Currently not reading location field in user input because it will be prepopulated later.

### CODE REVIEW

File: web-app/src/main/java/com/google/sps/data/Report.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.google.sps.data;
+
+public class Report {
+  private final String title;
+  private final double latitude;
+  private final double longitude;
+  private final long timestamp;
+  private final String incidentType;
+  private final String description;
+  /** TODO (Ashley): Add image. */
```
Comment (AshleyLTW) [inline web-app/src/main/java/com/google/sps/data/Report.java]: nit: AshleyLTW?

File: web-app/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.HashMap;
+import java.util.Date;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.sps.data.Report;
+
+/** Servlet that returns some example content. TODO: modify this file to handle comments data */
```
Comment (AshleyLTW) [inline web-app/src/main/java/com/google/sps/servlets/ReportServlet.java]: Should probably change this from the walkthrough comment

**Review by AshleyLTW — APPROVED** [review]

File: web-app/src/main/webapp/index.html [hunk 3]
```diff
@@ -73,7 +73,7 @@
         </div>
         <div class="row">
           <div class="two-thirds columns">
-            <button class="u-full-width" type="submit"
+            <button class="u-full-width" type="button"
```
Comment (darrnshn) [inline web-app/src/main/webapp/index.html]: do you need to put `type="button"` if it's already a button?

File: web-app/src/main/webapp/script.js [hunk 4]
```diff
@@ -59,8 +59,31 @@ function showMessageOnInfoWindow(message, position, map, infoWindow) {
...
+
+function reportFormToURLQuery() {
+  const PARAMS_FORM_MAP = new Map([
+    ['title-input', 'title'],
+    ['time-input', 'timestamp'],
+    ['category-input', 'incidentType'],
+    ['description-input', 'description'],
+  ]);
+
+  let searchParams = new URLSearchParams();
```
Comment (darrnshn) [inline web-app/src/main/webapp/script.js]: nit: const?

File: web-app/src/main/webapp/script.js [hunk 5]
```diff
@@ -59,8 +59,31 @@ function showMessageOnInfoWindow(message, position, map, infoWindow) {
...
+  for (const [formID, paramName] of PARAMS_FORM_MAP.entries()) {
+    const value = document.getElementById(formID).value;
+    searchParams.append(paramName, value);
+  }
+
+  return searchParams;
+}
+
+function postUserReport() {
+  const URLQuery = reportFormToURLQuery();
```
Comment (darrnshn) [inline web-app/src/main/webapp/script.js]: nit: naming should be `urlQuery`?

File: web-app/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    response.setContentType("text/html;");
+    response.getWriter().println("<h1>Hello world!</h1>");
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {    
+  
+    Entity reportEntity = new Entity("Report");
+
+    for (String paramName : PARAM_DEFAULT_MAP.keySet()) {
```
Comment (darrnshn) [inline web-app/src/main/java/com/google/sps/servlets/ReportServlet.java:53]: could we add some simple tests?

**Review by darrnshn — CHANGES_REQUESTED** [review]
Looks good, but hoping we can have some tests too!

File: web-app/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html;");
+    response.getWriter().println("<h1>Hello world!</h1>");
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {    
+
```
Comment (darrnshn) [inline web-app/src/main/java/com/google/sps/servlets/ReportServlet.java]: nit: remove blank line?

File: web-app/src/test/java/com/google/sps/servlets/ReportServletTest.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    when(request.getParameter("title")).thenReturn("Test");
+    when(request.getParameter("latitude")).thenReturn("1.11");
+    when(request.getParameter("longitude")).thenReturn("10.26");
+    when(request.getParameter("timestamp")).thenReturn("2017-06-01T08:30");
+    when(request.getParameter("incidentType")).thenReturn("Theft");
+    when(request.getParameter("description")).thenReturn("Sample request for testing");
+
+    Entity testReport = createReportEntity(request, response);
+    
+    Double latitude = 1.11;
```
Comment (darrnshn) [inline web-app/src/test/java/com/google/sps/servlets/ReportServletTest.java]: nit: you can probably just inline these into the test expectations

File: web-app/src/test/java/com/google/sps/servlets/ReportServletTest.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.appengine.api.datastore.Entity;
+
+@RunWith(JUnit4.class)
+public class ReportServletTest extends Mockito {
+  
+  @Test
+  public void testReportServlet() throws IOException {
```
Comment (darrnshn) [inline web-app/src/test/java/com/google/sps/servlets/ReportServletTest.java:23]: nit: maybe a test with invalid parameters,and check that we get an exception?

**Review by darrnshn — APPROVED** [review]

**Pull request merged**